### PR TITLE
feat: Create GraphQL API endpoints for email service credentials

### LINKED ISSUE #1129: [TAB-437] Mail Delivery capabilities.
Create a mail service in tabby-ee that supports reading SMTP information from database and use it to send email.

This service should enable us to send emails for Tabby, such as invitation links and password reset notifications.

- [X] Table design for mail delivery settings.
- [X] GraphQL API for setting / reading smtp informations.
- [X] MailService in tabby webserver.

<sub>From [SyncLinear.com](https://synclinear.com) | [TAB-437](https://linear.app/tabbyml/issue/TAB-437/mail-delivery-capabilities)</sub>

### PR DESCRIPTION
**Pull request by boxbeam**
Closes #1129

### CODE REVIEW

File: ee/tabby-webserver/src/schema/dao.rs [hunk 1]
```diff
@@ -54,3 +57,13 @@ impl From<GithubOAuthCredentialDAO> for OAuthCredential {
         }
     }
 }
+
+impl From<EmailServiceCredentialDAO> for EmailServiceCredential {
+    fn from(value: EmailServiceCredentialDAO) -> Self {
```
Comment (wsxiaoys) [inline ee/tabby-webserver/src/schema/dao.rs:62]: can we add a flag to email service credential dao for active flag? so we can disable email service as well

File: ee/tabby-webserver/src/schema/email_service_credential.rs [hunk 2]
```diff
@@ -0,0 +1,16 @@
...
+#[derive(GraphQLObject)]
+pub struct EmailServiceCredential {
+    pub smtp_username: String,
+    pub smtp_password: String,
+    pub smtp_server: String,
+}
+
+#[async_trait]
+pub trait EmailServiceCredentialService: Send + Sync {
+    async fn get_email_service_credential(&self) -> Result<Option<EmailServiceCredential>>;
```
Comment (wsxiaoys) [inline ee/tabby-webserver/src/schema/email_service_credential.rs:14]: do not return password in read interface

**Review by wsxiaoys — CHANGES_REQUESTED** [review]

File: ee/tabby-webserver/src/schema/dao.rs (hunk 1 above)
Comment (boxbeam) [reply ee/tabby-webserver/src/schema/dao.rs:62]: Since it's stored as an Option, do we just want to use its presence or absence to indicate whether it's enabled? Or would we ever need to store the credentials while it's inactive?

Comment (wsxiaoys) [reply ee/tabby-webserver/src/schema/dao.rs:62]: good point, I'm fine to add an delete interface to make it represents "inactive" (as we don't really need to store the credentials)

Once you done with this, could you also refactor https://github.com/TabbyML/tabby/blob/main/ee/tabby-db/src/github_oauth_credential.rs#L31 to making "delete" representing inactive?

File: ee/tabby-webserver/src/schema/dao.rs [hunk 3]
```diff
@@ -1,5 +1,8 @@
-use tabby_db::{GithubOAuthCredentialDAO, InvitationDAO, JobRunDAO, UserDAO};
+use tabby_db::{
+    EmailServiceCredentialDAO, GithubOAuthCredentialDAO, InvitationDAO, JobRunDAO, UserDAO,
+};
 
+use super::email_service_credential::EmailServiceCredential;
```
Comment (wsxiaoys) [inline ee/tabby-webserver/src/schema/dao.rs:5]:
```suggestion
use crate::schema::email_service_credential::EmailServiceCredential;
```

File: ee/tabby-webserver/src/schema/email_service_credential.rs [hunk 4]
```diff
@@ -0,0 +1,18 @@
...
+#[derive(GraphQLObject)]
+pub struct EmailServiceCredential {
+    pub smtp_username: String,
+    pub smtp_server: String,
+}
+
+#[async_trait]
+pub trait EmailServiceCredentialService: Send + Sync {
+    async fn get_email_service_credential(&self) -> Result<Option<EmailServiceCredential>>;
+    async fn update_email_service_credential(&self, creds: EmailServiceCredentialDAO)
```
Comment (wsxiaoys) [inline ee/tabby-webserver/src/schema/email_service_credential.rs:15]:
```suggestion
    async fn update_email_service_credential(&self, smtp_username: &str, smtp_password: &str, smtp_server: &str)
```

Comment (wsxiaoys) [reply ee/tabby-webserver/src/schema/email_service_credential.rs:15]: do not reuse DAO for update input (in future, DAO might add fields like created_at / updated_at / ...)

File: ee/tabby-webserver/src/schema/email_service_credential.rs [hunk 5]
```diff
@@ -12,7 +12,11 @@ pub struct EmailServiceCredential {
 #[async_trait]
 pub trait EmailServiceCredentialService: Send + Sync {
     async fn get_email_service_credential(&self) -> Result<Option<EmailServiceCredential>>;
-    async fn update_email_service_credential(&self, creds: EmailServiceCredentialDAO)
-        -> Result<()>;
+    async fn update_email_service_credential(
+        &self,
+        smtp_username: String,
+        smtp_password: String,
```
Comment (wsxiaoys) [inline ee/tabby-webserver/src/schema/email_service_credential.rs:17]:
```suggestion
        smtp_password: Optional<String>,
```

Since we do not pass password to client side, we need to support updates without password, so users could change SMTP server / protocol without having to retype the password

File: ee/tabby-webserver/src/schema/mod.rs [hunk 6]
```diff
@@ -243,6 +246,17 @@ impl Query {
         )))
     }
 
+    async fn email_service_credential(
+        ctx: &Context,
+    ) -> FieldResult<Option<EmailServiceCredential>> {
```
Comment (wsxiaoys) [inline ee/tabby-webserver/src/schema/mod.rs:251]: use Result

File: ee/tabby-webserver/src/schema/mod.rs [hunk 7]
```diff
@@ -378,6 +392,20 @@ impl Mutation {
             "Only admin is able to update oauth credential",
         ))
     }
+
+    async fn update_email_service_credential(
+        ctx: &Context,
+        smtp_username: String,
+        smtp_password: Option<String>,
+        smtp_server: String,
+    ) -> FieldResult<bool> {
```
Comment (wsxiaoys) [inline ee/tabby-webserver/src/schema/mod.rs:401]: ditto

File: ee/tabby-webserver/src/service/email_service_credential.rs [hunk 8]
```diff
@@ -0,0 +1,42 @@
...
+        Ok(creds.map(Into::into))
+    }
+
+    async fn update_email_service_credential(
+        &self,
+        smtp_username: String,
+        smtp_password: Option<String>,
+        smtp_server: String,
+    ) -> Result<()> {
+        let smtp_password = match smtp_password {
```
Comment (wsxiaoys) [inline ee/tabby-webserver/src/service/email_service_credential.rs:22]: move the optional logic inside of tabby-db, instead of here.

File: ee/tabby-webserver/src/service/email_service_credential.rs [hunk 9]
```diff
@@ -0,0 +1,42 @@
...
+            Some(pass) => pass,
+            None => {
+                let Some(creds) = self.read_email_service_credential().await? else {
+                    return Err(anyhow!("No existing credential to update"));
+                };
+                creds.smtp_password
+            }
+        };
+        self.update_email_service_credential(EmailServiceCredentialDAO {
```
Comment (wsxiaoys) [inline ee/tabby-webserver/src/service/email_service_credential.rs:31]: do not reuse DAO struct for updating, pass smtp_username / smtp_password / smtp_server directly.

File: ee/tabby-db/src/email_service_credential.rs [hunk 10]
```diff
@@ -40,8 +40,22 @@ impl DbConn {
...
         &self,
-        creds: EmailServiceCredentialDAO,
+        smtp_username: String,
+        smtp_password: Option<String>,
+        smtp_server: String,
     ) -> Result<()> {
+        let smtp_password = match smtp_password {
+            Some(pass) => pass,
+            None => {
+                self.read_email_service_credential()
```
Comment (wsxiaoys) [inline ee/tabby-db/src/email_service_credential.rs:50]: would it make sense to execute this query inside of call block to make it a transaction? so there's a gurantee of atomic for write-after-read

File: ee/tabby-db/src/email_service_credential.rs [hunk 11]
```diff
@@ -44,21 +44,15 @@ impl DbConn {
...
-                        anyhow!("No existing SMTP credentials, password must be specified")
-                    })?
-                    .smtp_password
-            }
-        };
-
         Ok(self
             .conn
             .call(move |c| {
+                 let smtp_password = match smtp_password {
```
Comment (wsxiaoys) [inline ee/tabby-db/src/email_service_credential.rs:50]: I don't think this ensures the transaction semantic.

See https://docs.rs/rusqlite/latest/rusqlite/struct.Transaction.html for an example

File: ee/tabby-db/src/email_service_credential.rs [hunk 12]
```diff
@@ -47,13 +47,14 @@ impl DbConn {
         Ok(self
             .conn
             .call(move |c| {
+                let transaction = c.transaction()?;
                  let smtp_password = match smtp_password {
                     Some(pass) => pass,
                     None => {
```
Comment (wsxiaoys) [inline ee/tabby-db/src/email_service_credential.rs:53]: please add a unit test for this

Comment (boxbeam) [reply ee/tabby-db/src/email_service_credential.rs:53]: If you meant add a unit test to ensure the transaction is atomic, I'm not sure how to do that. I added another test case covering updating everything other than the password, which wasn't covered before. Let me know if I misunderstood.

Comment (wsxiaoys) [reply ee/tabby-db/src/email_service_credential.rs:53]: right, i mean to add a test for calling `update_email_service_credential` with `smtp_password=None`, saw you've added it.

File: ee/tabby-db/src/repositories.rs
Comment (wsxiaoys) [inline ee/tabby-db/src/repositories.rs:1]: seems to me the change to this file is irrelevant to the PR? wanna create a separate PR for it?

File: ee/tabby-webserver/src/service/email_service_credential.rs [hunk 13]
```diff
@@ -0,0 +1,29 @@
+use anyhow::Result;
+use async_trait::async_trait;
+use tabby_db::DbConn;
+
+use crate::schema::email_service_credential::{
+    EmailServiceCredential, EmailServiceCredentialService,
+};
+
+#[async_trait]
+impl EmailServiceCredentialService for DbConn {
```
Comment (wsxiaoys) [inline ee/tabby-webserver/src/service/email_service_credential.rs:10]:
```suggestion
impl EmailService for DbConn {
```

Comment (wsxiaoys) [reply ee/tabby-webserver/src/service/email_service_credential.rs:10]: Please updating the file name correspondingly

File: ee/tabby-db/src/repositories.rs
Comment (boxbeam) [reply ee/tabby-db/src/repositories.rs:1]: I can do that

Comment (boxbeam) [reply ee/tabby-db/src/repositories.rs:1]: Actually, it doesn't compile without this, is it fine to leave it in? I believe I had to make this change while resolving merge conflicts.

File: ee/tabby-webserver/src/schema/mod.rs [hunk 14]
```diff
@@ -38,6 +42,7 @@ pub trait ServiceLocator: Send + Sync {
     fn logger(&self) -> Arc<dyn RawEventLogger>;
     fn job(&self) -> Arc<dyn JobService>;
     fn repository(&self) -> Arc<dyn RepositoryService>;
+    fn email_service_credential(&self) -> Arc<dyn EmailService>;
```
Comment (wsxiaoys) [inline ee/tabby-webserver/src/schema/mod.rs:45]:
```suggestion
    fn email(&self) -> Arc<dyn EmailService>;
```

File: ee/tabby-webserver/src/schema/mod.rs [hunk 15]
```diff
@@ -432,6 +446,20 @@ impl Mutation {
...
+        smtp_password: Option<String>,
+        smtp_server: String,
+    ) -> Result<bool> {
+        let _service = ctx.locator.email_service_credential();
+        ctx.locator
+            .email_service_credential()
+            .update_email_service_credential(smtp_username, smtp_password, smtp_server)
+            .await?;
+        Ok(true)
+    }
```
Comment (wsxiaoys) [inline ee/tabby-webserver/src/schema/mod.rs:462]: add mutation for for delete?

File: ee/tabby-db/src/email_service_credential.rs
Comment (wsxiaoys) [inline ee/tabby-db/src/email_service_credential.rs:1]: seems to me everything named `email_service_credentials` are better being named `email_settings`, including the table names in tabby-db.

I'm fine to merge it as is for now, but please follow up with a refactor PR w/ renamings.

**Review by wsxiaoys — COMMENTED** [review]
Otherwise LGTM

File: ee/tabby-db/migrations/07-email-service-credentials/down.sql
Comment (wsxiaoys) [inline ee/tabby-db/migrations/07-email-service-credentials/down.sql:1]: please also update dir name of migrations

File: ee/tabby-db/src/email_settings.rs [hunk 16]
```diff
@@ -0,0 +1,117 @@
+use anyhow::Result;
+use rusqlite::{named_params, OptionalExtension};
+
+use crate::DbConn;
+
+const EMAIL_CREDENTIAL_ROW_ID: i32 = 1;
+
+#[derive(Debug, PartialEq)]
+pub struct EmailServiceCredentialDAO {
```
Comment (wsxiaoys) [inline ee/tabby-db/src/email_settings.rs:9]:
```suggestion
pub struct EmailSettingDAO {
```

File: ee/tabby-db/migrations/07-email-service-credentials/down.sql [hunk 17]
```diff
@@ -1 +1 @@
-DROP TABLE email_service_credential;
+DROP TABLE email_settings;
```
Comment (wsxiaoys) [inline ee/tabby-db/migrations/07-email-service-credentials/down.sql:1]:
```suggestion
DROP TABLE email_setting;
```

Since we'll only have one row of email setting

File: ee/tabby-db/src/email_settings.rs [hunk 18]
```diff
@@ -0,0 +1,117 @@
...
+        Self {
+            smtp_username,
+            smtp_password,
+            smtp_server,
+        }
+    }
+}
+
+impl DbConn {
+    pub async fn read_email_settings(&self) -> Result<Option<EmailServiceCredentialDAO>> {
```
Comment (wsxiaoys) [inline ee/tabby-db/src/email_settings.rs:26]: `settings` -> `setting`, similar to others.

File: ee/tabby-db/src/email_settings.rs
Comment (wsxiaoys) [inline ee/tabby-db/src/email_settings.rs:1]: `email_setting.rs`

**Review by wsxiaoys — CHANGES_REQUESTED** [review]

File: ee/tabby-webserver/src/schema/mod.rs [hunk 19]
```diff
@@ -432,6 +442,24 @@ impl Mutation {
...
+        smtp_server: String,
+    ) -> Result<bool> {
+        ctx.locator
+            .email_setting()
+            .update_email_setting(smtp_username, smtp_password, smtp_server)
+            .await?;
+        Ok(true)
+    }
+
+    async fn delete_email_setting(ctx: &Context) -> Result<bool> {
```
Comment (wsxiaoys) [inline ee/tabby-webserver/src/schema/mod.rs:459]:
```suggestion
    async fn delete_email_setting(ctx: &Context) -> Result<()> {
```

**Review by wsxiaoys — COMMENTED** [review]
Otherwise LGTM

File: ee/tabby-webserver/src/schema/mod.rs (hunk 19 above)
Comment (boxbeam) [reply ee/tabby-webserver/src/schema/mod.rs:459]: This is not allowed, `()` does not implement a trait required in order to return it from a graphql endpoint. A `bool` which is always true is used everywhere else for this.
